### PR TITLE
OAuth2 filter: Optimize HMAC encoding

### LINKED ISSUE #28728: OAuth2: Optimization: Use only base64 or only hex encoding for the HMAC cookie
The OAuth2 filter creates a HMAC and saves it in a cookie called "OauthHMAC". For this cookie, the HMAC is first hex-encoded, and then base64-encoded.

It's not necessary to use both encodings. Only hex encoding or only base64 encoding is sufficient, and the cookie value will be shorter.

*Example:*

Hex encoding + Base64 encoding (current implementation):
```
ZWU2ZDIwOWU0ODQ1Mjk5ZDczODA3MWI3ODQxN2I3YTg0MDg5YmE1MzA4ZTk5MjBhNmY0M2E3ODBlMjZiNDMzMw==
```

Hex encoding only:
```
ee6d209e4845299d738071b78417b7a84089ba5308e9920a6f43a780e26b4333
```

Base64 encoding only:
```
7m0gnkhFKZ1zgHG3hBe3qECJulMI6ZIKb0OngOJrQzM=
```

### PR DESCRIPTION
**Pull request by Ronnie-personal**
Commit Message: remove base64 encoding from OAuth2 HMAC cookie, apply hex encoding only.
Additional Description:
Risk Level: Low
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
Fixes #28728
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

**Comment by Ronnie-personal** [comment]
Hi @htuch 
Would you mind helping me with a question? Is it possible to debug test/extensions/filters/http/oauth2/filter_test.cc locally?

I have configured the Visual Studio Code Dev Container based on the guide at https://github.com/envoyproxy/envoy/blob/main/tools/vscode/README.md.
But I don't see run test/debug test button in the IDE.

Thanks.

**Comment by htuch** [comment]
I'm not super familiar with the VS Code flow, I tend to use vim :) That said, you can definitely run tests via Bazel locally. Maybe you are missing a Bazel plugin for VS Code?

**Comment by Ronnie-personal** [comment]
@htuch 
Can you please help me to understand the status of the pipeline?  There 22 jobs displayed as 'Waiting for status to be reported'. I have tried to trigger the CI for couple of times, but those jobs are still not kicked off.

Thanks.

**Comment by htuch** [comment]
Since you are a new contributor, CI requires a maintainer to kick off the CI.

**Comment by yanavlasov** [comment]
This change is user visible, i.e. it modifies Envoy's behavior in a way that can potentially break some users. In this case we use runtime features to allow switching back to the old behavior in case some Envoy operator was impacted by this change. 

Please add a new runtime feature to switch back to the old behavior. See PR #27974 for example on how to do it.

/wait

**Comment by Ronnie-personal** [comment]
> This change is user visible, i.e. it modifies Envoy's behavior in a way that can potentially break some users. In this case we use runtime features to allow switching back to the old behavior in case some Envoy operator was impacted by this change.
> 
> Please add a new runtime feature to switch back to the old behavior. See PR #27974 for example on how to do it.
> 
> /wait

Thanks for the review! I will look into the runtime features.

**Comment by derekargueta** [comment]
Since the runtime guard defaults to true, wouldn't this force every user to re-authenticate when a new Envoy build with this change gets deployed?

**Comment by derekargueta** [comment]
two other thoughts here:
1) I'm wondering if there oughtta be a migrate path - ability to allocate tokens in the new format while still respecting tokens of the old format. That way operators can safely roll out the new envoy version with no surprises and after a reasonable amount of time has passed we can safely axe the b64+hex

2) I'm curious if there's a reason for choosing only hex rather than only base64? Base64 seems to yield shorter values per the github issue.

**Comment by Ronnie-personal** [comment]
> Since the runtime guard defaults to true, wouldn't this force every Oauth user to re-authenticate when an operator deploys a new Envoy build with this change?

Great question. I'm pretty new to this open-source project, so I don't know the exact impact. Do we know how Oauth token is  validated?

**Comment by Ronnie-personal** [comment]
> two other thoughts here:
> 
>     1. I'm wondering if there oughtta be a migrate path - ability to allocate tokens in the new format while still respecting tokens of the old format. That way operators can safely roll out the new envoy version with no surprises and after a reasonable amount of time has passed we can safely axe the b64+hex
> 
>     2. I'm curious if there's a reason for choosing only hex rather than only base64? Base64 seems to yield shorter values per the github issue.

Thanks for the review. 
1. I will look into respecting the token of the old format.
2. No, there is not specific reason for choosing hex, the requirement does not call out the exact option. I just choose one of the two options. I definitely can refactor the code to use base64.

File: source/extensions/filters/http/oauth2/filter.cc [hunk 1]
```diff
@@ -145,9 +145,13 @@ std::string encodeHmac(const std::vector<uint8_t>& secret, absl::string_view hos
   auto& crypto_util = Envoy::Common::Crypto::UtilitySingleton::get();
   const auto hmac_payload =
       absl::StrJoin({host, expires, token, id_token, refresh_token}, HmacPayloadSeparator);
-  std::string encoded_hmac;
-  absl::Base64Escape(Hex::encode(crypto_util.getSha256Hmac(secret, hmac_payload)), &encoded_hmac);
-  return encoded_hmac;
+  if (Runtime::runtimeFeatureEnabled("envoy.reloadable_features.hmac_hex_encoding_only")) {
+    return Hex::encode(crypto_util.getSha256Hmac(secret, hmac_payload));
```
Comment (kanurag94) [inline source/extensions/filters/http/oauth2/filter.cc:149]: Hex will take two characters for each byte while Base64 takes 4 characters for every 3 bytes, so it should be more efficient to just use base64. Wdyt?

Comment (Ronnie-personal) [reply source/extensions/filters/http/oauth2/filter.cc:149]: I agree, we can go with Base64.

Comment (Ronnie-personal) [reply source/extensions/filters/http/oauth2/filter.cc:149]: Do you have any guidance on 'respecting the token of the old format'? 
I can update the OAuth2 filter validator to accept both 'base64' and 'hex+base64'. However, how do we disable this behavior when the user no longer needs this backward compatibility?

Comment (kanurag94) [reply source/extensions/filters/http/oauth2/filter.cc:149]: I agree accepting the old token is a good idea. What we may be able to do is to accept the base64+hex token and `set-cookie` with the new token. This way when this backward compatibility is not required and becomes the default behaviour we can remove the runtime guard safely. There may be better ideas, but this seems simple enough.

**Comment by yanavlasov** [comment]
@Ronnie-personal new tests are failing: https://source.cloud.google.com/results/invocations/66045611-6d71-461b-b5e7-fc2792cbf56e

**Comment by Ronnie-personal** [comment]
> @Ronnie-personal new tests are failing: https://source.cloud.google.com/results/invocations/66045611-6d71-461b-b5e7-fc2792cbf56e

Thanks, I'm fixing it right now.

**Comment by Ronnie-personal** [comment]
The pipeline failed with '##[error]We stopped hearing from agent i-060b64c3735e1bae4. Verify the agent machine is running and has a healthy network connection. Anything that terminates an....'

Shall I rerun the pipeline?

**Comment by yanavlasov** [comment]
@Ronnie-personal  I've restarted CI, but there merge conflict that needs resolving.

**Comment by Ronnie-personal** [comment]
> /wait



> @Ronnie-personal I've restarted CI, but there merge conflict that needs resolving.

Thanks! I will resolve the conflict.

**Comment by Ronnie-personal** [comment]
I will investigate further on the pipeline failure. This PR does not touch 'substitution_formatter_test.cc'.

Here is the error:
test/common/formatter/substitution_formatter_test.cc:934:38: error: use of undeclared identifier 'request_headers'
              upstream_format.format(request_headers, response_headers, response_trailers,

**Review by yanavlasov — APPROVED** [review]

**Pull request merged**